src/devices/cddadevice.cpp

```diff
@@ -36,24 +40,53 @@ CddaDevice::CddaDevice(const QUrl& url, DeviceLister* lister,
           SLOT(SongsLoaded(SongList)));
   connect(this, SIGNAL(SongsDiscovered(SongList)), model_,
           SLOT(SongsDiscovered(SongList)));
+  //   connect(&cd_device_, SIGNAL(DiscChanged()), SLOT(DiscChangeDetected()));
+  cdio_ = cdio_open(url_.path().toLocal8Bit().constData(), DRIVER_DEVICE);
+  if (!cdio_) {
+    throw DeviceError(url.toString(),
+                      "Cannot open device: cdio_open returned nullptr");
+  }
+  connect(&disc_changed_timer_, SIGNAL(timeout()), SLOT(CheckDiscChanged()));
+  WatchForDiscChanges(watch_for_disc_changes);
 }
 
-CddaDevice::~CddaDevice() {}
-
-void CddaDevice::Init() {
-  song_count_ = 0;  // Reset song count, in case it was already set
-  cdda_song_loader_.LoadSongs();
+CddaDevice::~CddaDevice() {
+  Q_ASSERT(cdio_);
+  cdio_destroy(cdio_);
 }
 
-void CddaDevice::Refresh() {
-  if (!cdda_song_loader_.HasChanged()) {
-    return;
-  }
-  Init();
+void CddaDevice::Init() { LoadSongs(); }
+
+CddaSongLoader* CddaDevice::loader() { return &cdda_song_loader_; }
+
+CdIo_t* CddaDevice::raw_cdio() { return cdio_; }
```

Comment:
Have you found cdio documentation anywhere that discusses thread safety? It doesn't seem to be very well documented in that regard. It may be necessary to examine the code.

Reply:
cdio does not seem to state anything about thread safety, but it seems to be blocking/synchronised via the system calls it does in the background.